### PR TITLE
Feature: Support for external ATS for parachutes

### PR DESCRIPTION
**Pull request by dakejahl**
**Describe problem solved by this pull request**
Support for external automatic trigger systems for parachutes (ATS) as required by ASTM F3322-18. This serves to terminate the flight while the ATS handles deploying of the chute.

i.e: [SATS-MINI](https://fruitychutes.com/help_for_parachutes/sats-mini-ats-quick-start-guide.htm)

**Describe your solution**
Added uORB publishing to the `pwm_input` module. Added parameters to enable the ATS and configure the PWM trigger threshold . Added another status check in `Commander::FailureDetector` to monitor the PWM input on AUX5.

**Describe possible alternatives**
We could not support this, but I think it is useful. Perhaps @eengelgau can chime in here.

**Test data / coverage**
Tested on the bench successfully.

FYI
@eengelgau
@dagar

### CODE REVIEW

**Comment by dakejahl** [comment]
Dang, flash overflow. Looks like I need to take care of CDev usage right meow.

**Comment by dakejahl** [comment]
Completely removed CDev usage from `pwm_input` and `ll40ls`, moved to just uORB. Cleaned up dead code, fixed the `pwmin_info` command, and refactored out lidar lite specific code to the `ll40ls` driver. 

Tested successfully on the bench, if anyone has a LidarLite it would be good to verify the PWM interface still works as expected.

File: src/drivers/pwm_input/pwm_input.cpp [hunk 1]
```diff
@@ -314,160 +214,43 @@ void PWMIN::_timer_init(void)
...
-	} else {
-		printf("count=%u period=%u width=%u\n",
-		       (unsigned)_pulses_captured,
-		       (unsigned)_last_period,
-		       (unsigned)_last_width);
-	}
+	PX4_INFO("count=%u period=%u width=%u\n",
+		 (unsigned)_pulses_captured,
+		 (unsigned)_last_period,
+		 (unsigned)_last_width);
```
Comment (mcsauder) [inline src/drivers/pwm_input/pwm_input.cpp]: Looks good so far @dakejahl.  Using `static_cast<unsigned int>(var_name)` will allow compile time checking where the current c-style cast will not.

**Comment by eengelgau** [comment]
Hi Everyone, this is Gene Engelgau with Fruity Chutes.  We manufacture parachute recovery systems for drones. Last year (2018) I was involved with the ASTM standards committee that created the specification for what it means to have a certified parachute recovery system for a drone for the purposes of flight over people, or for other high-risk use.  The standard in the F.38 section is ASTM F3322-18 and is available here for purchase:

  https://webstore.ansi.org/Standards/ASTM/ASTMF332218

The FAA (and other similar government agencies in other countries) are now referring people to this standard who want waivers for flight over people. In that specification there are some mandatory requirements for several key items that also affect the autopilot.  First, the standard says that there shall be a Autonomous Trigger System (ATS) that will detect the drone failure to trigger the parachute.  We make this device here that satisfies this requirement:

 https://fruitychutes.com/uav_rpv_drone_recovery_parachutes/sentinel-automatic-trigger-system.htm

Below is the F3322-18 clause that defines this:

3.1.4 autonomous triggering system, ATS, n—device or components independent from any ﬂight critical system of the sUA that will detect and initiate parachute deployment upon detection of a critical failure of the sUA in ﬂight. 

The ATS has it's own IMU and other sensors that can detect the drone falling or flipping (out of control motion that should otherwise never happen).  In this case the ATS will trigger the parachute.  
The ATS also will generate another signal that is sent to the Autopilot to stop the rotors from spinning. This is called the flight termination system.  The best and most reliable way to do this is to have the autopilot accept a PWM pulse width (or other ways can be done as well like logic level) that when asserted will cause all the rotors to immediately stop. Here is the paragraph from the standard that refers to the flight termination (stop the rotors)

3.1.16 ﬂight termination system, FTS, n—device or components that will disable the propulsion system of the sUA. 

Below are several paragraphs that makes the assertions regarding these: 

5.8.4 To prevent propellers from injuring people on the ground after a deployment the PRS shall also include a Flight Termination System (FTS). 

5.8.4.1 The FTS shall activate at the time of or before parachute deployment is initiated.

----
Having this feature added to the Autopilot is the best way to do this.  

Contact me if people have questions - Gene Engelgau, Fruity Chutes Inc.

PS - A company in California is already flight testing this feature using our ATS system (SATS-MINI).

File: src/modules/commander/failure_detector/FailureDetector.cpp [hunk 2]
```diff
@@ -132,3 +136,36 @@ FailureDetector::updateAttitudeStatus()
...
+{
+	bool updated(false);
+	pwm_input_s pwm_input;
+
+	if (_sub_pwm_input.update(&pwm_input)) {
+
+		bool ats_trigger_status = false;
+		uint32_t pulse_width = pwm_input.pulse_width;
+
+		if (pulse_width >= (uint32_t)_param_fd_ext_ats_trig.get()) {
```
Comment (bkueng) [inline src/modules/commander/failure_detector/FailureDetector.cpp]: Might be worth filtering unreasonably long pulses as well (e.g. > 3ms)

File: src/modules/commander/failure_detector/failure_detector_params.c [hunk 3]
```diff
@@ -103,3 +103,25 @@ PARAM_DEFINE_FLOAT(FD_FAIL_R_TTRI, 0.3);
...
+/**
+ * The PWM threshold from external automatic trigger system for engaging failsafe.
+ *
+ * External ATS is required by ASTM F3322-18.
+ *
+ * @boolean
+ * @reboot_required true
+ * @group Commander
+ */
+PARAM_DEFINE_INT32(FD_EXT_ATS_TRIG, 1990);
```
Comment (bkueng) [inline src/modules/commander/failure_detector/failure_detector_params.c]: The margin here seems too low. I'd use something like 1900.

File: src/modules/commander/failure_detector/FailureDetector.cpp [hunk 4]
```diff
@@ -132,3 +136,36 @@ FailureDetector::updateAttitudeStatus()
...
+			ats_trigger_status = true;
+		}
+
+		hrt_abstime time_now = hrt_absolute_time();
+
+		// Update hysteresis
+		_ext_ats_failure_hysteresis.set_hysteresis_time_from(false, (hrt_abstime)(1e6f * 0.1f)); // 5 consecutive pulses at 50hz
```
[the remark this answers is not in the document]
Comment (dakejahl) [reply src/modules/commander/failure_detector/FailureDetector.cpp]: Which header do I need to include?

**Comment by dakejahl** [comment]
@bkueng I wasn't sure how to make ModuleBase work with the timer ISR. Since the ISR needs to access functions from PWMIN, and currently there is no method to share the ModuleBase class pointer to functions outside of the class.

**Comment by dakejahl** [comment]
Another thing worth mentioning. Daniel suggested using the `px4::work_queue`, however I unsure if interrupts + wq scheduling is going to cause issues.

i.e: interrupt fires, sets member data "status, period, pulse_width", and calls `ScheduleNow`. If another interrupt fires during the execution of `Run` there could be a data race. Also requires member variable access from a static member function, which is annoying.

File: src/modules/commander/failure_detector/FailureDetector.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/commander/failure_detector/FailureDetector.cpp]: using namespace time_literals;

**Comment by dagar** [comment]
> Another thing worth mentioning. Daniel suggested using the px4::work_queue for the uORB publication, however it seems like interrupts + wq scheduling might cause issues.

Either way is fine here, but longer term I want to see if we can get the majority of work (anything involving floating point or ORB) out of ISRs.

**Comment by dakejahl** [comment]
Blocked by v2 flash limit

File: src/drivers/pwm_input/pwm_input.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	uint32_t _error_count {};
+	uint32_t _pulses_captured {};
+	uint32_t _last_period {};
+	uint32_t _last_width {};
+
+	bool _timer_started {};
+
+	pwm_input_s _pwm {};
+
+	uORB::PublicationData<pwm_input_s> _pwm_input_pub { ORB_ID(pwm_input) };
```
Comment (dagar) [inline src/drivers/pwm_input/pwm_input.h]:
```suggestion
	uORB::PublicationData<pwm_input_s> _pwm_input_pub{ORB_ID(pwm_input)};
```

File: src/modules/commander/failure_detector/failure_detector_params.c [hunk 6]
```diff
@@ -103,3 +103,26 @@ PARAM_DEFINE_FLOAT(FD_FAIL_R_TTRI, 0.3);
...
+ * @group Commander
+ */
+PARAM_DEFINE_INT32(FD_EXT_ATS_EN, 0);
+
+/**
+ * The PWM threshold from external automatic trigger system for engaging failsafe.
+ *
+ * External ATS is required by ASTM F3322-18.
+ *
+ * @boolean
```
Comment (hamishwillee) [inline src/modules/commander/failure_detector/failure_detector_params.c]: Doesn't look very boolean to me.

**Comment by hamishwillee** [comment]
FYI This will require (at least) an update to http://docs.px4.io/master/en/peripherals/parachute.html. From what I can tell this is just a signal on a particular port (MAIN5/AUX5) that tells the system that parachute triggering has happened? (or is it "must happen" - and PX4 does the triggering?). 
How does this interact with the existing [flight termination behaviour](http://docs.px4.io/master/en/advanced_config/flight_termination.html)?

@bkueng Triggering off fixed ports has always been a bit of a pain for camera - ie we're always running into particular frames that use the ports already etc. Is there no way to make these things more configurable?

**Comment by dakejahl** [comment]
> From what I can tell this is just a signal on a particular port (MAIN5/AUX5) that tells the system that parachute triggering has happened?

Yes that is correct. This PR simply engages flight termination based on that signal.

**Comment by hamishwillee** [comment]
> Yes that is correct. This PR simply engages flight termination based on that signal.

Thank you! Is everything else respected? i.e. the circuit breaker etc?

**Comment by dakejahl** [comment]
> Thank you! Is everything else respected? i.e. the circuit breaker etc?
Yes, this behaves in exactly the same way as exceeding the roll/pitch limits

**Comment by dakejahl** [comment]
@dagar Are we OK to merge?

**Pull request merged**